### DISCUSSION TITLE
How do I convert Data to CSR format while still containing all nodes

### DISCUSSION THREAD

**Discussion by ProfDoof** [pyg-team/pytorch_geometric#7305] [Q&A] [answered]
https://github.com/pyg-team/pytorch_geometric/blob/fb1d855d5e3d12092ffaa8d456e3a46252cf9723/torch_geometric/data/graph_store.py#L277-L281

TL;DR. The CSR format that this is converted to does not have a long enough rowptr tensor to represent graphs that have isolated nodes where such nodes have a higher index than any of the source nodes for the edge indices that you define. This means that the random_walk implementation in both pyg_lib and torch_cluster fails. With that being said, is there a way I can get the CSR representation without losing these nodes?

More explanation:

So, I have been getting this fun runtime error in my implementation of node2vec that I have been writing (I have my reasons, but they aren't relevant to this story). It was a CUDA Runtime Error saying I was going out of bounds and accessing memory I should not be. However, as far as I could tell, I was not going out of bounds. Then, I looked deeper, and it turns out that given a graph like the following
![test](https://user-images.githubusercontent.com/33268432/236557726-c013f332-3faa-465d-8c05-85719ce23f6d.png)

the CSR representation is as follows
```py
rowptr=tensor([0, 1, 1, 2, 3], device='cuda:0')

col=tensor([3, 1, 2], device='cuda:0')
```

This means that we are losing information about the graph because the CSR representation, which includes nodes 4 and 5, should look like

```py
rowptr=tensor([0, 1, 1, 2, 3, 3, 3], device='cuda:0')

col=tensor([3, 1, 2], device='cuda:0')
```

This feels like it should be addressed in an issue if it's a problem, but maybe I'm missing something, so I want to ask first.

**Comment by grantnedwards** [comment]
you have isolated nodes in your graph that are not being included in the CSR representation, causing issues in your node2vec implementation. To preserve these isolated nodes, you can modify the code that converts the COO format to CSR.

```python
row = torch.tensor([0, 2, 3], dtype=torch.long)
col = torch.tensor([3, 1, 2], dtype=torch.long)
num_nodes = 6  # Optional, include the isolated nodes

rowptr, col = coo_to_csr(row, col, num_nodes)
```

you could try this function here to handle COO to CSR - above is how you can call and assign it.

```python
def coo_to_csr(row, col, num_nodes=None):
    if num_nodes is None:
        num_nodes = int(row.max()) + 1

    row, perm = index_sort(row, max_value=num_nodes)
    col = col[perm]
    
    rowptr = torch.zeros(num_nodes + 1, dtype=row.dtype, device=row.device)
    rowptr[1:] = (row[:-1] != row[1:]).to(rowptr.dtype)
    rowptr = torch.cumsum(rowptr, dim=0)

    return rowptr, col
```

**Reply by ProfDoof** [reply]
```py
from torch_geometric.data import Data, EdgeAttr
from torch_geometric.data.graph_store import EdgeLayout
from torch_geometric.utils import index_sort
from torch_geometric.utils.sparse import index2ptr, ptr2index
import copy

def my_edge_to_layout(data: Data, attr: EdgeAttr, layout: EdgeLayout, store: bool = False):
    (row, col), perm = data.get_edge_index(attr), None

    if layout == EdgeLayout.COO:  # COO output requested:
        if attr.layout == EdgeLayout.CSR:  # CSR->COO
            row = ptr2index(row)
        elif attr.layout == EdgeLayout.CSC:  # CSC->COO
            col = ptr2index(col)

    elif layout == EdgeLayout.CSR:  # CSR output requested:
        if attr.layout == EdgeLayout.CSC:  # CSC->COO
            col = ptr2index(col)

        if attr.layout != EdgeLayout.CSR:  # COO->CSR
            if attr.size: 
                num_rows = attr.size[0]
            elif data.num_nodes:
                num_rows = data.num_nodes
            else:
                num_rows = int(row.max()) + 1
            
            row, perm = index_sort(row, max_value=num_rows)
            col = col[perm]
            row = index2ptr(row, num_rows)
    
    # CSC output requested:
    elif attr.layout == EdgeLayout.CSR:  # CSR->COO
            row = ptr2index(row)

    elif attr.layout != EdgeLayout.CSC:  # COO->CSC
        num_cols = attr.size[1] if attr.size else int(col.max()) + 1
        if not attr.is_sorted:  # Not sorted by destination.
            col, perm = index_sort(col, max_value=num_cols)
            row = row[perm]
        col = index2ptr(col, num_cols)

    if attr.layout != layout and store:
        attr = copy.copy(attr)
        attr.layout = layout
        if perm is not None:
            attr.is_sorted = False
        data.put_edge_index((row, col), attr)

    return row, col, perm

Data._edge_to_layout = my_edge_to_layout
```

Based on your code I was able to write the above monkey patch.

**Reply by ProfDoof** [reply]
This is very fragile, and I can't recommend doing it, but it should work for what I'm doing.

**Reply by rusty1s** [reply]
Do you have a reproducible example for this? Where does that happen exactly? Are you operating on a custom `GraphStore`? If yes, try to pass the size of nodes to the `GraphStore` as well.

**Reply by ProfDoof** [reply]
I'm not working with a custom graph store. It's just the base `Data` type.

I can make a reproducible example in a bit.

**Reply by ProfDoof** [reply]
@rusty1s Here's the actual and expected behavior from me:

```
Actual behavior: data.csr()=({None: tensor([0, 2])}, {None: tensor([1, 0])}, {None: tensor([0, 1])})
Expected behavior: data.csr()=({None: tensor([0, 2, 2, 2, 2, 2, 2])}, {None: tensor([1, 0])}, {None: tensor([0, 1])})
```

Here's the code I used to generate the previous outputs:

```py
from torch_geometric.data import Data
import torch
from local import my_edge_to_layout # This is a local folder with an __init__.py that contains the function I posted above.

data = Data(torch.tensor([0, 1, 2, 3, 4, 5]), torch.tensor([[0, 0], [1, 0]]))
data.num_nodes = 6
print(f'Actual behavior: {data.csr()=}')

data.coo()
Data._edge_to_layout = my_edge_to_layout
print(f'Expected behavior: {data.csr()=}')
```

**Reply by ProfDoof** [reply]
The function I wrote isn't a complete solution but it works for my use case. I assume y'all would have to do more.

**Reply by ProfDoof** [reply]
Oh, and my version of pytorch_geometric is 2.3.0, I'm using torch version 2.0.0

**Reply by rusty1s** [reply, marked as answer]
Thanks for the example, helped a lot. I fixed it in https://github.com/pyg-team/pytorch_geometric/pull/7316. Sorry for any inconvenience.

**Reply by ProfDoof** [reply]
Np, thanks for the fix.

**Reply by ProfDoof** [reply]
@rusty1s dumb question, when will that fix be integrated into the conda package?

**Reply by rusty1s** [reply]
This will be integrated once we release a new version (we target June right now). Not sure if this issue warrants a hotfix to make it available faster. After all, you can also install PyG from master to have this fix integrated on your end.

**Reply by ProfDoof** [reply]
That's fine, I just wanted to check. I am trying to install it from main using Conda but having lots of fun issues. 😂 We'll see if I can get it working correctly.

**Reply by ProfDoof** [reply]
I got it working; fun fact, you have to change the source from the torch_geometric pypi bundle to the pyg_nightly pypi bundle. And you have to add a dependency on flit for some reason. I'm not sure why.

**Reply by rusty1s** [reply]
Can you clarify? this sounds like a monster hack, and I am not sure why any of this would be needed :(

**Reply by ProfDoof** [reply]
The meta.yaml you have in the Conda directory is pointed at torch_geometric on pypi. So I had to point it at the nightly version instead. Then, when I did that it started breaking saying something about flit_core not being available so I added a dependency on flit to the host object in the meta.yaml. but after that, it was pretty simple to build.
